timeboost-sequencer/src/decrypt.rs

```diff
@@ -570,115 +639,173 @@ impl Worker {
-            ));
+        if !self.dkg_completed.contains(&committee_id) {
+            trace!(node = %self.label, %committee_id, "local dkg incomplete");
+            return Ok(());
         }
 
-        let Some(committee) = self.committees.get(committee_id) else {
-            return Err(DecrypterError::NoCommittee(committee_id));
-        };
-
-        committee
+        self.key_stores
+            .read()
+            .get(committee_id)
+            .ok_or(DecrypterError::NoCommittee(committee_id))?
+            .committee()
             .get_index(&src)
-            .ok_or_else(|| DecrypterError::UnknownKey(src))?;
+            .ok_or(DecrypterError::UnknownKey(src))?;
 
         let subset = match self.dkg_tracker.get(&committee_id) {
-            Some(acc) if acc.completed() => acc.bundles(),
+            Some(acc) => acc.bundles(),
             _ => {
                 return Err(DecrypterError::Dkg(format!(
-                    "DKG incomplete for requested committee: {committee_id}",
+                    "no subset available for committee: {committee_id}",
                 )));
             }
         };
 
-        let response = SubsetResponse {
-            committee_id,
-            subset: Subset::new(committee_id, subset.to_vec()),
-        };
+        let subset = DkgSubset::new(committee_id, subset.to_vec());
+        let response = SubsetResponse::new(committee_id, subset);
+
         self.net
             .unicast(
                 src,
-                0, // Minimal round number since API requires it
-                serialize(&Protocol::GetResponse(response))?,
+                self.oldest_cached_round().u64(),
+                serialize(&Protocol::DkgResponse(response))?,
             )
             .await
             .map_err(|e| DecrypterError::End(e.into()))?;
 
         Ok(())
     }
 
-    /// A get response for DKG subset has been received.
-    async fn on_get_response(&mut self, src: PublicKey, res: SubsetResponse) -> Result<()> {
-        let SubsetResponse {
-            committee_id,
-            subset,
-        } = res;
-        trace!(node = %self.label, from=%src, %committee_id, "received get_response");
+    /// A response for DKG subset has been received.
+    async fn on_dkg_response_msg(&mut self, src: PublicKey, res: SubsetResponse) -> Result<()> {
+        trace!(node = %self.label, from=%src, %res.committee_id, "received dkg response");
+        if res.committee_id != self.current {
+            trace!(node = %self.label, from=%src, %res.committee_id, "not current committee");
+            return Ok(());
+        }
 
-        let DkgState::Pending(ref mut subsets) = self.dkg_state else {
-            trace!("received get_response but not in a recovering state");
+        let WorkerState::DkgPending(ref mut subsets) = self.state else {
+            trace!(node = %self.label, from=%src, %res.committee_id, "not in a pending state");
```

Comment:
this didn't account for the catchup during resharing (epoch 2+)
`DkgRequest()` could fire for committee in epoch 2+, and the receiving node could be in a state of `AwaitingHandover()` (namely the late-joining nodes in epoch 2)

Reply:
yes, i was asking about the non-initial DKG epoch catchup, 
so basically we don't have that (neither do we have test for that) atm. Probably need to address this in future PR.